Speed up first asset load by encoding asset manifest in binary rather than JSON

andrewkolos opened the pull request:
## Context
When loading an asset for the first time, the framework has to load a file called the asset manifest, which contains information about assets available to the app (for example, where to find higher-resolution [variants of images]). This file is in the JSON format.

Reading the file from storage, decoding into UTF, and parsing it as JSON is generally slower than we'd like, especially since this is will all likely happen on first paint (since many apps include assets on their first screens).

The asset manifest (`AssetManifest.json`) is created at build time.

## Change

The asset manifest will now be in a binary format rather than JSON. In particular, we'll use the [`standard_message_codec` package] for encoding the manifest at build time and decoding it at runtime.  In addition, for decoding the binary message, we'll lazily put off typecasting until we read individual keys--avoiding upfront typecasting for all entries. This will achieve a significant speedup for first asset load, especially on less powerful devices.

Here are the results from running the [`load_and_parse_large_asset_manifest` benchmark](https://github.com/flutter/flutter/blob/master/dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart#L35) on a Galaxy A02S before and after the change:

* Before: 9608ms, 9539ms, 9508ms (average 9551.7ms)
* After: 1955ms, 1876ms, 1881ms (average 1904ms)

In apps with large asset manifests, such as money (g3), running on lower-spec devices, this should hopefully result in a speed up of at least one frame when loading.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[variants of images]: https://docs.flutter.dev/development/ui/assets-and-images#asset-variants
[`standard_message_codec` package]: https://pub.dev/packages/standard_message_codec

andrewkolos committed:
initial attempt to use binary encoding for assetmanifest

andrewkolos committed:
working

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
trim trailing whitespace

andrewkolos committed:
update asset decoding test

andrewkolos committed:
Delete generated_plugins.cmake

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -228,10 +236,46 @@ abstract class CachingAssetBundle extends AssetBundle {
+
+    Completer<dynamic>? completer;
+    Future<dynamic>? result;
+    load(key).then<void>((ByteData value) {
```
This could probably use `compute` to move the parsing to a separate thread

andrewkolos replied:
I tried that with https://github.com/flutter/flutter/pull/113637/files#diff-1b6c749abe1192b78f5ddab5a43a95e382b6a7d9dc4fd675559d7e4b8800c8daR245-R251, but I get the following when trying to run the [benchmark](https://github.com/andrewkolos/flutter/blob/3739c09b256d2e34d09cf204dc3c413be17bcdf1/dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart#L16).

```
E/flutter ( 2582): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Binding has not yet been initialized.
E/flutter ( 2582): The "instance" getter on the ServicesBinding binding mixin is only available once that binding has been initialized.
E/flutter ( 2582): Typically, this is done by calling "WidgetsFlutterBinding.ensureInitialized()" or "runApp()" (the latter calls the former). Typically this call is done in the "void main()" method. The "ensureInitialized" method is idempotent; calling it multiple times is not harmful. After calling that method, the "instance" getter will return the binding.
E/flutter ( 2582): In a test, one can call "TestWidgetsFlutterBinding.ensureInitialized()" as the first line in the test's "main()" method to initialize the binding.
E/flutter ( 2582): If ServicesBinding is a custom binding mixin, there must also be a custom binding class, like WidgetsFlutterBinding, but that mixes in the selected binding, and that is the class that must be constructed before using the "instance" getter.
E/flutter ( 2582): #0      BindingBase.checkInstance.<anonymous closure> (package:flutter/src/foundation/binding.dart:287:9)
E/flutter ( 2582): #1      BindingBase.checkInstance (package:flutter/src/foundation/binding.dart:369:6)
E/flutter ( 2582): #2      ServicesBinding.instance (package:flutter/src/services/binding.dart:55:54)
E/flutter ( 2582): #3      PlatformAssetBundle.load (package:flutter/src/services/asset_bundle.dart:308:31)
E/flutter ( 2582): #4      CachingAssetBundle.thinkies (package:flutter/src/services/asset_bundle.dart:275:34)
E/flutter ( 2582): #5      _IsolateConfiguration.applyAndTime.<anonymous closure> (package:flutter/src/foundation/_isolates_io.dart:108:21)
E/flutter ( 2582): #6      Timeline.timeSync (dart:developer/timeline.dart:166:22)
E/flutter ( 2582): #7      _IsolateConfiguration.applyAndTime (package:flutter/src/foundation/_isolates_io.dart:106:21)
E/flutter ( 2582): #8      _spawn (package:flutter/src/foundation/_isolates_io.dart:127:67)
E/flutter ( 2582): #9      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:300:17)
E/flutter ( 2582): #10     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
E/flutter ( 2582): 
Reloaded 1 of 846 libraries in 1,117ms (compile: 27 ms, reload: 418 ms, reassemble: 403 ms).
Reloaded 451 of 846 libraries in 6,153ms (compile: 3433 ms, reload: 1512 ms, reassemble: 249 ms).
```

I tried adding `TestWidgetsFlutterBinding.ensureInitialized();` as the first statement in the test, but that remedy the issue anything. Have you seen this before?

jonahwilliams replied:
Don't load the data in the compute isolate, pass the data that is already loaded through the args. Just do the decode in the isolate.

andrewkolos replied:
I found that doing this results in a significant slowdown ([code](https://github.com/andrewkolos/flutter/blob/7cd4ca08f40cc2d387572211aa1f969ad1a8ddb0/packages/flutter/lib/src/services/asset_bundle.dart#L246-L260)). The decoding only takes 2-3ms, perhaps the overhead from using another isolate is too much.

andrewkolos replied:
Disregard. I just saw your new comment below.

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -457,17 +466,17 @@ class ManifestAssetBundle implements AssetBundle {
+  void _setIfChanged(String key, DevFSContent content, AssetKind assetKind) {
+    final DevFSContent? oldContent = entries[key];
+
+    // TODO(andrewkolos): Why is the "IfChanged" part important?
```
if the object is not backed by a file or something, then the tool will assume that new entiries need to be updated and these will get copied to the device on each hot reload/restart, which is wasteful

andrewkolos replied:
Where/when would this assumption happen? Is there some caching mechanism that compares values by reference?

andrewkolos committed:
try to use compute

andrewkolos committed:
experiment with compute

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -242,13 +243,23 @@ abstract class CachingAssetBundle extends AssetBundle {
-      return result;
-    }, null);
+    // ~10,000 ms
+    // final ByteData data = await load(key);
```
Ahh yeah, we'd have to copy the data back to the compute isolate to do the decode since ByteData is mutable. For now its probably best to keep it on the main thread.

andrewkolos committed:
remove json version of asset manifest

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
remove padding from asset manifest

andrewkolos committed:
fix buffer length issue

andrewkolos committed:
fix messed up method name

andrewkolos committed:
trim trailing whitespace

andrewkolos committed:
remove trailing whitespace

andrewkolos commented:
Right now, I've somehow made first manifest load slower rather than faster (according to the benchmark), so I'm trying to figure that out.

jonahwilliams reviewed (commented):
Something to be aware of is that you don't really want to mix `async` and synchronous future. The point of the latter is that in case you are fully cached you return the sync future so that you don't wait an even turn

jonahwilliams reviewed (commented):
On a Pixel 6 in release mode, the new method is taking about 800ms and the old method takes about 1600ms. How are you testing this?

andrewkolos commented:
Oh yeah, my apologies for leaving that out. I'm running the new benchmark, setting [the number of iterations to 1](https://github.com/flutter/flutter/pull/113637/files#diff-450953f593a0e60cf541ea5a948606e8e9a15d09a512ed4749541aad0af05756L13). On master, the benchmark completes in about ~125ms, and on this branch it runs in ~140ms. This is on a Samsung Galaxy S9. IIRC from Friday, the new method is faster, but only with a higher number of iterations.

jonahwilliams commented:
Its possible that with 1 iteration you're mostly measuring the time to load the asset. Since both are about the same size, perhaps you should move that out of the benchmark?

jonahwilliams commented:
Are you loading the bytes outside of the benchmark in the JSON case: https://github.com/flutter/flutter/blob/master/dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart#L23

andrewkolos commented:
@jonahwilliams. LUCI is unhappy with me including the test asset manifest (money_asset_manifest.bin) because it's not UTF8. https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8799564695396497569/+/u/Analyze/analyze/test_stdout. I see two ways forward, but I'm not thrilled by either of them:

1. Move the file to some other repo as the error message suggests.
2. Include the asset manifest file in some other format (e.g. go back to using JSON) and then, when setting up the test, read it and re-encode it in the standard message format.

Do you have any thoughts or other options I am not aware of?

jonahwilliams commented:
You should add an exception to the analysis test - we have some other examples of doing it. The intention of that check is to prevent folks from checking in large binary assets accidentally

see dev/bots/test.dart

andrewkolos committed:
run flutter update-packages --force-upgrade

andrewkolos committed:
add money_asset_manifest.bin to _legacyBinaries

jonahwilliams reviewed (commented):
I would plan on leaving in the code that generates and parse the JSON version of the asset manifest. We need to leave both in order to support a soft transition between the two formats. This gives us time to update any dependencies that parse the manifest (google fonts possibly) as well as the google3 specific tooling that has its own manifest generation path

andrewkolos committed:
Revert "run flutter update-packages --force-upgrade"

This reverts commit 23decc2ad1d3ffb5e96acfa0cc35644a6671c5e5.

andrewkolos committed:
restore JSON version of asset manifest

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
restore iteration count

andrewkolos committed:
remove pointless bundle.clear call

andrewkolos commented:
Talked with @jonahwilliams yesterday and wanted to summarize my takeaways for my own record. Also, while I'm happy with the current approach in the change, I can't help but (re)ask some questions.

This change will deprecate the `AssetBundle.loadStructuredData` API, replacing it with a more general form that operates directly on ByteData, `loadStructuredDataBinary`. If we are OK with deprecating `loadStructuredData`, I can't help but think about how `loadStructuredData` might be currently used and whether or not it should have ever been exposed to begin with. As I understand it, the primary intent of `loadString` and `loadStructuredData` is to provide built-in caching of string and structured data (e.g. JSON) assets respectively (`loadString` also conveniently provides UTF-8 decoding). The idea to me, then, is if there are disparate codebases that access the same (text/structured data) asset, then the app will benefit from caching. I still am not sure about the idea (how often are disparate codebases requesting text/JSON assets?), but changing it now could risk unexpected performance degradation for not much gain, so I don't think it's worth worrying about now.

Caching aside, outside of the tool itself, is there a known use case of `loadStructuredData` aside from reading the asset manifest (e.g. google fonts for figuring it out if it can avoid loading fonts over the network because they are already available)?
* If there is no such use case, would the google fonts use case be better as something we explicitly support with additional APIs (e.g. an API that determines whether or not a font with a particular name is available)? I ask this because the idea of external code loading and inspecting `AssetManifest.json` directly feels like reverse-engineering. If we plan to deprecate `loadStructuredData` and thus break that means of accessing `AssetManifest.json`, then, I feel like, while we are at it, we may as well improve the API here so folks don't have to reach into flutter internals like `AssetManifest.json` (or `AssetManifest.bin`).
* If there is such a use case, then do we really want to deprecate `loadStructuredData`? If folks, for example, want to include JSON (or similar) files as assets, then why inconvenience them by having them switch to `loadStructuredDataBinary` and then prepend UTF8 decoding into their parser function? Why not just keep both APIs around?

andrewkolos committed:
normalize API

jonahwilliams commented on dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart:
```diff
@@ -2,32 +2,30 @@
 
 void main() async {
-  assert(false, "Don't run benchmarks in debug mode! Use 'flutter run --release'.");
+  //assert(false, "Don't run benchmarks in debug mode! Use 'flutter run --release'.");
```
Add this back

jonahwilliams commented on dev/benchmarks/microbenchmarks/lib/foundation/decode_and_parse_asset_manifest.dart:
```diff
@@ -2,32 +2,30 @@
 import '../common.dart';
 
-const int _kNumIterations = 1000;
+const int _kNumIterations = 1;
```
Maybe 10-ish iterations?

andrewkolos replied:
Oops, this was in error. I should keep it at 1000 to maintain parity with the current version of the benchmark, no?

jonahwilliams replied:
Yes, good idea

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -328,27 +327,23 @@ class AssetImage extends AssetBundleImageProvider {
+  static Future<Map<String, List<String>>> decodeAssetManifest(ByteData data) {
+    final Object? decoded = const StandardMessageCodec().decodeMessage(data);
+    final Map<dynamic, dynamic> typed = decoded! as Map<dynamic, dynamic>;
+    final Map<String, List<String>> result = <String, List<String>>{
+      for (final MapEntry<dynamic, dynamic> entry in typed.entries)
+        entry.key as String: (entry.value as List<dynamic>).cast<String>(),
     };
-    // TODO(ianh): convert that data structure to the right types.
-    return SynchronousFuture<Map<String, List<String>>?>(parsedManifest);
+    return SynchronousFuture<Map<String, List<String>>>(result);
```
Nothing in this API is async, I would remove the Future usage and just return the object. Also, if you can avoid the casting, that would speed this up a bit. We're already breaking the API, so you can leave the new one as untyped.

andrewkolos replied:
> Nothing in this API is async, I would remove the Future usage and just return the object.

Since `loadStructuredData` takes an asynchronous `parser`, I want to preserve that property in `loadStructuredDataBinary` to make it easier for customers to transition to. In my latest commit, I changed the type of the `parser` parameter to be a function that returns a `FutureOr<T>`. Let me know if this isn't idiomatic use of `FutureOr`--if so, we could introduce a `-Sync` version of the method.

> Also, if you can avoid the casting, that would speed this up a bit. We're already breaking the API, so you can leave the new one as untyped.

Okay, I removed all the `List.cast` calls since they definitely add up for large manifests. `decodeAssetManifest` seems to be twice as fast now. I left in the cast to `Map<dynamic, dynamic>` since we'll need to cast anyway to avoid a lint violation (indexing into a `dynamic`).

jonahwilliams replied:
Sounds good. You might be able to further reduce casting by making the select variant accept an Object? and defer calling toString until you actually need to Uri parse

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -8,8 +8,8 @@ import 'dart:io';
 import 'dart:ui' as ui;
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
```
Don't import services.dart, import the individual libraries you need from services, since this library is itself part of services

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -228,16 +244,40 @@ abstract class CachingAssetBundle extends AssetBundle {
+  /// subsequent calls will be a [SynchronousFuture], which resolves its
+  /// callback synchronously.
+  @override
+  Future<T> loadStructuredDataBinary<T>(String key, Future<T> Function(ByteData data) decoder) async {
```
Because this method is `async` you will still need to wait an extra event loop even if it is cached. Write it more like loadStructuredData above

andrewkolos commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -154,11 +161,19 @@ class NetworkAssetBundle extends AssetBundle {
   /// fetched.
   @override
   Future<T> loadStructuredData<T>(String key, Future<T> Function(String value) parser) async {
-    assert(key != null);
```
`key` isn't nullable, so is it even possible for `key` to be null here?

jonahwilliams replied:
Yes, if you have not migrated to null safety yet. We're only going to remove these once it is possible not to run without null safety

andrewkolos committed:
rewrite loadStructuredDataBinary to avoid event loop if data is in cache

andrewkolos committed:
provide base implementation for loadStructuredDataBinary

This is to avoid breaking folks that currently extend AssetBundle.

andrewkolos committed:
remove asynchrony from decodeAssetManifest and weaken types

andrewkolos committed:
make import for ServicesBinding more specific

andrewkolos committed:
fix test

andrewkolos committed:
fix Failed image loads in debug mode test

andrewkolos committed:
adjust verifyNoBinaries

andrewkolos committed:
remove unnecessary await

andrewkolos committed:
fix tests that inspect bundle.entries

andrewkolos committed:
fix more tests

andrewkolos commented:
Heya @Hixie. Per [this line of code in analyze.dart](https://github.com/flutter/flutter/blob/35cbbb54afb1da8727bc24a24410359ceecaea22/dev/bots/analyze.dart#L1540), I wanted to ask you about adding a new file to `_legacyBinaries`. For the purpose of benchmarking, this PR adds an [asset manifest file](https://github.com/flutter/flutter/pull/113637/files#diff-01db0a514b49084b631d07bf4979df3d947a67a63a9b0e7bc89b6d6bc28f67fe) (which is now binary rather than JSON with this change). This file is a little under 120KB. Do you have any reservations about me adding this file?

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -9,7 +9,7 @@ import 'dart:ui' as ui;
 import 'package:flutter/foundation.dart';
 
-import 'binding.dart';
+import './binding.dart';
```
```suggestion
import 'binding.dart';
```

andrewkolos committed:
adjust binding import

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -622,39 +635,47 @@ class ManifestAssetBundle implements AssetBundle {
+    Map<String, List<String>> assetManifest
+  ) {
+    const StandardMessageCodec codec = StandardMessageCodec();
+    final ByteData result = codec.encodeMessage(assetManifest)!;
```
This format is the same as the old format, whereas I believe the design document lists a few options for expanding this to support more fields (locale/light dark mode) by making the entries objects.

andrewkolos replied:
Yeah. Yesterday, I realized I totally skipped that part. Added a note about that in the PR description. I'll work on it.

Hixie commented:
We shouldn't add a 120KB binary file. Binary files end up being very expensive for every user of Flutter forever, because of how we distribute our code. We have other repositories for that kind of thing. In this case though it seems like it would be even better to generate the manifest on the fly for the test? These manifest files should never be checked in to any repos in normal production environments right?

(Side note, we should presumably add this file to our default `.gitignore`?)

jonahwilliams commented:
I was sort of hoping we could get away with it since _most_ of the data is actually utf8 encoded strings and in theory should have the same diff behavior as a large file filled with strings. If not, thats OK - we can change the app under test to generate the synthetic manfiest by encoding it as JSON and then converting it to SMC on the fly.

We don't need to add this file to the .gitignore because its normally generated as a side effect of the build and not user visable.

jonahwilliams commented:
actually @andrewkolos you should probably be sure to _not_ delete the original JSON file, otherwise we won't be able to modify the binary SMC file without writing a separate helper program to parse it again

andrewkolos committed:
Revert "adjust verifyNoBinaries"

This reverts commit 35cbbb54afb1da8727bc24a24410359ceecaea22.

andrewkolos committed:
(broken) have asset manifest use lists of dynamic objects for values

andrewkolos committed:
fix previous commit

andrewkolos committed:
clean up previous commit

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -30,6 +30,9 @@ const String kFontManifestJson = 'FontManifest.json';
 final RegExp _assetVariantDirectoryRegExp = RegExp(r'/?(\d+(\.\d*)?)x$');
 
+// We assume the main asset is designed for a device pixel ratio of 1.0
+const double _defaultResolution = 1.0;
```
constants should be prefixed with `k` according to the style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#begin-global-constant-names-with-prefix-k

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -1,17 +1,17 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
```
FYI, you'll need to fix these lints before you submit

andrewkolos replied:
Oh wow I must have accidentally added these via VSCode

jonahwilliams replied:
These are still here FYI

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -284,17 +281,21 @@ class AssetImage extends AssetBundleImageProvider {
-        final double chosenScale = _parseScale(chosenName);
+          candidateVariants,
+        );
+        // ignore: avoid_dynamic_calls
```
Don't ignore dynamic calls

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -284,17 +281,21 @@ class AssetImage extends AssetBundleImageProvider {
           bundle: chosenBundle,
-          name: chosenName,
+          // ignore: avoid_dynamic_calls
+          name: chosenVariant['asset'] as String,
```
Same here

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -328,35 +329,26 @@ class AssetImage extends AssetBundleImageProvider {
   }
 
-  String? _chooseVariant(String main, ImageConfiguration config, List<String>? candidates) {
+  dynamic _chooseVariant(String main, ImageConfiguration config, List<dynamic>? candidates) {
```
_chooseVariant should return the string that is needed to load the asset in question. Any additional facets used to select this asset should be passed into this function (i.e. brightness)

Hixie commented:
> I was sort of hoping we could get away with it since _most_ of the data is actually utf8 encoded strings and in theory should have the same diff behavior as a large file filled with strings.

It's not at all clear to me how well that would work, binary git deltification is a bit of a mystery to me. But in general, any large file is always a problem regardless, just because it's large.


> If not, thats OK - we can change the app under test to generate the synthetic manfiest by encoding it as JSON and then converting it to SMC on the fly.

How was the synthetic data created in the first place?


> We don't need to add this file to the .gitignore because it's normally generated as a side effect of the build and not user visible.

If it's already `.gitignore`d by virtue of being put somewhere that's already in `.gitignore` then that's great.

Hixie commented:
Can we document this binary format?

jonahwilliams commented:
Yes, we should document the format (at least starting here in the asset code), including which keys we support today, and which ones we intend to add in the future

andrewkolos commented:
> How was the synthetic data created in the first place?

I took the existing asset manifest of an internal app (money) and wrote an ad-hoc program to obfuscate it (though I don't believe the data to be sensitive in any way) and convert it to the binary format we will use going forward.

andrewkolos committed:
centralize asset manifest parsing logic within image resolution logic

(also removes the main asset from it's list of variants in order to save storage and parsing time)

andrewkolos committed:
remove nullability of asset manifest

is there ever case where we would explicitly write null to AssetManifest.bin?

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
remove abbreviation

andrewkolos committed:
revert update to binary allow list

andrewkolos committed:
resolve library_private_types_in_public_api lint

andrewkolos committed:
update asset_bundle_package_test.dart tests

jonahwilliams reviewed (commented):
Is this ready for another review?

andrewkolos committed:
update more tests

andrewkolos commented:
I'm updating more tests now. I should do a pass myself, but if you are eager to review then go for it!

Question: For the tests involving asset manifests, should we temporarily keep the tests involving `AssetManifest.json` and just make duplicates for the binary version? I feel like that's the right thing to do, but wanted to run it by you before I make the effort.

Also FYI I updated the manifest format so that each asset is no longer a variant of itself--to reduce parsing time. 
That is `{"images/foo.jpg":["images/foo.jpg"]` is now `{"images/foo.jpg":[]}`.

The default dpr image resolution logic should remain unchanged (i.e. if targeting a dpr if 1.0, we should still pick the main/default version of the asset). This should be fine unless we want room to one day make the keys only logical/symbolic ones (i.e. they no longer represent actual file locations).

jonahwilliams commented:
> Question: For the tests involving asset manifests, should we temporarily keep the tests involving AssetManifest.json and just make duplicates for the binary version? I feel like that's the right thing to do (given that the json is being deprecated--not removed), but wanted to run it by you before I make the effort.

What would be the value of the benchmark then? If the old JSON benchmark regressed would you drop what you are doing to fix it? Or would we say "well, we're moving away from this anyway, and we know its slow 🤷‍♂️ ". (Probably the latter).

Everything else SGTM.

andrewkolos commented:
> > Question: For the tests involving asset manifests, should we temporarily keep the tests involving AssetManifest.json and just make duplicates for the binary version? I feel like that's the right thing to do (given that the json is being deprecated--not removed), but wanted to run it by you before I make the effort.
> 
> What would be the value of the benchmark then? If the old JSON benchmark regressed would you drop what you are doing to fix it? Or would we say "well, we're moving away from this anyway, and we know its slow 🤷‍♂️ ". (Probably the latter).
> 
> Everything else SGTM.

I wasn't referring to the benchmark. I'm referring to tests like those found in `asset_bundle_package_test.dart` and `asset_bundle_variant_test.dart` that make sure the manifests we are generating contain the correct data.

jonahwilliams commented:
Ah yes, in that case definitely keep them around. We'll delete them eventually

andrewkolos committed:
restore AssetManifest.json tests

andrewkolos committed:
fix tests in asset_bundle_test.dart

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
fix image_resolution tests

andrewkolos committed:
fix tests in asset_bundle_test

andrewkolos commented:
@jonahwilliams I'm looking at this test I broke: https://github.com/flutter/flutter/blob/56c4e3cdd8a11ba373979e05f48836d1440e7d25/packages/flutter/test/widgets/image_resolution_test.dart#L262-L271

This test was added with https://github.com/flutter/flutter/pull/11564. The idea behind it makes enough sense, but when I try to load an image that _only_ has variants, the result is an error (on stable). 
<details>
<summary>repro (macos)</summary>

[Repro link](https://github.com/andrewkolos/flutter_issue_repros/tree/no-main-asset/simple_asset_issues)
![2022-11-03_18-24-08 (1)](https://user-images.githubusercontent.com/9027551/199865180-6976f1f5-162d-4d96-82dd-dc24632f236c.gif)
</details>

To your knowledge, is it possible for an image to only have variants?

andrewkolos committed:
fix tests in image_resolution_test.dart

jonahwilliams commented:
I've not tried. its possible that this was implemented and then broke and some point in the future.

andrewkolos committed:
nits

andrewkolos committed:
nit

andrewkolos commented:
Ready for review again. I held off on running `flutter upgrade-packages` since that would muddy up the diff. Let me know if some of the diffs of the test files prove to be a bit much. I prioritized being able to quickly remove tests that are no longer needed once `AssetManifest.json` is gone for good.

andrewkolos committed:
add deprecation message and rename to loadStructuredBinaryData

loadStructuredBinaryData is a clearer name that's more consistent with how most developers label the data formats we are dealing with. The previous name shares a common prefix with loadStructuredData, increasing discoverability, but this discoverability is less necessary given we are deprecating loadStructuredData.

andrewkolos committed:
make deprecation message slightly less ambiguous

andrewkolos committed:
Revert "make deprecation message slightly less ambiguous"

This reverts commit 8f58c27f32284d49399dad7bf9a7a129541896b0.

andrewkolos committed:
Revert "add deprecation message and rename to loadStructuredBinaryData"

This reverts commit 24231f1d53f3fc946178057132f5704711f8291f.

andrewkolos committed:
add documentation of the asset manifest format

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
broken

andrewkolos commented:
@jonahwilliams Could I borrow your experience with the `Completer` API? My latest change adds logic for falling back `AssetManifest.json` in the case `AssetManifest.bin` is missing ([code](https://github.com/andrewkolos/flutter/blob/17b5b9fb210056fcf9a426e6fc8a3b64b7b89881/packages/flutter/lib/src/painting/image_resolution.dart#L289)). However, it seems to only work the first time an image is loaded (e.g. load a route with an image on it, pop the route, nav into it again and the image won't show up).

andrewkolos committed:
simplify asset manifest loading logic

andrewkolos commented:
> Could I borrow your experience with the `Completer` API? My latest change adds logic for falling back `AssetManifest.json` in the case `AssetManifest.bin` is missing ([code](https://github.com/andrewkolos/flutter/blob/17b5b9fb210056fcf9a426e6fc8a3b64b7b89881/packages/flutter/lib/src/painting/image_resolution.dart#L289)). However, it seems to only work the first time an image is loaded (e.g. load a route with an image on it, pop the route, nav into it again and the image won't show up).

The issue appears to stem from use of `onError`. If I comment out [the onError chain for falling back to the json asset manifest](https://github.com/andrewkolos/flutter/blob/abc800bddf20f59181f793a9d65a83f43463daa2/packages/flutter/lib/src/painting/image_resolution.dart#L294), it appears to work again.

andrewkolos committed:
fix future chain

andrewkolos committed:
nit

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -430,4 +416,152 @@ class AssetImage extends AssetBundleImageProvider {
+    return _AssetManifestBin(data as Map<dynamic, dynamic>);
+  }
+
+  late final Map<dynamic, dynamic> _data;
```
Does this need to be late?

andrewkolos committed:
perf nit

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

andrewkolos committed:
remove unnecessary late

jonahwilliams commented on packages/flutter/test/widgets/image_resolution_test.dart:
```diff
@@ -259,46 +257,21 @@ void main() {
     // the 1.0x asset.
 
-    const String manifest = '''
+    final Map<dynamic, dynamic> manifest = json.decode('''
```
Here and elsewhere, `Map<dynamic, dynamic>` is the same runtime type as `Map<Object?, Object?>`, however the latter will prevent accidentally making dynamic calls on either the keys/objects retrieved from the map. I would change the type signatures where possible

christopherfujino replied:
This is part of our style guide too https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var-and-dynamic

jonahwilliams commented on packages/flutter_tools/test/general.shard/asset_bundle_variant_test.dart:
```diff
@@ -15,209 +16,418 @@ import 'package:flutter_tools/src/base/user_messages.dart';
-'''
-name: test
-dependencies:
+  // TODO(andrewkolos): Delete this group once we stop producing AssetManifest.json
```
For TODOs with a task associated like this, Its good practice to file a cleanup bug and link it here

andrewkolos replied:
Good idea. I'll get on it

andrewkolos replied:
Added a link to https://github.com/flutter/flutter/issues/114913 in the TODO

andrewkolos committed:
refactor _LegacyAssetManifest

andrewkolos committed:
Merge branch 'add-binary-encoding-for-asset-manifest' of https://github.com/andrewkolos/flutter into add-binary-encoding-for-asset-manifest

andrewkolos committed:
formatting nit

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -548,6 +529,26 @@ class _LegacyAssetManifest implements _AssetManifest {
   }
+
+  static double _parseScale(String mainAsset, String variant) {
+    /** The legacy asset manifest includes the main asset within its variant list. */
```
use `//` for comments and `///` for doc comments

andrewkolos committed:
nit

andrewkolos committed:
remove uses of dynamic

andrewkolos committed:
add cleanup issue

andrewkolos committed:
remove accidental lint ignores

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
correct linux analyze lints

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
flutter update-packages --force-upgrade

andrewkolos committed:
upgrade assertions to use matchers

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
run "flutter update-packages --force-upgrade"

jonahwilliams reviewed (approved):
LGTM

jonahwilliams commented:
@andrewkolos I think the google testing shard is stuck. You'll need to investigate and/or ask for help on what is going on

jonahwilliams commented:
google testing is stuck, i've asked @andrewkolos to click submit

jonahwilliams commented:
actually turns out its not stuck, we just messed up applying the fix!

jonahwilliams commented on packages/flutter/lib/src/painting/image_resolution.dart:
```diff
@@ -284,18 +285,29 @@ class AssetImage extends AssetBundleImageProvider {
+    // TODO(andrewkolos): Once google3 and google-fonts-flutter are migrated
+    // away from using AssetManifest.json, remove all references to it.
+    // See https://github.com/flutter/flutter/issues/114913.
+    chosenBundle
```
Based on my read of the presubmit failures, it looks like we're not catching the error that propagates when loadStructuredBinaryData fails to lookup an asset, causing an unhandled async exception.

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -225,16 +251,53 @@ abstract class CachingAssetBundle extends AssetBundle {
+
+    Completer<T>? completer;
+    Future<T>? result;
+    load(key).then<T>(parser).then<void>((T value) {
+      result = SynchronousFuture<T>(value);
+      _structuredDataBinaryCache[key] = result!;
+      if (completer != null) {
+        completer.complete(value);
+      }
+    });
+    if (result != null) {
+      return result!;
+    }
+    completer = Completer<T>();
```
The first call to load will throw an exception if key does not map to any asset. You'll need to add an error listener to this one and correcty add an error to the completer.

andrewkolos committed:
fix AssetManifest.json fallback

andrewkolos committed:
i dont understand how async works in dart

andrewkolos commented:
@jonahwilliams See [previous commit](https://github.com/flutter/flutter/pull/113637/commits/eed94ec255add02bc8c3bf8e5f03000323710ffb).

https://github.com/flutter/flutter/blob/eed94ec255add02bc8c3bf8e5f03000323710ffb/packages/flutter/lib/src/painting/image_resolution.dart#L294-L302

The commented out `onError` doesn't work as I expect (see test failures in `image_resolution_test.dart`), yet the "identity then" with an `onError` argument that I'm currently employing does 😕

andrewkolos committed:
Update asset_bundle.dart

andrewkolos committed:
i am oh so confused

andrewkolos committed:
go back to using identity then with onError arg

jonahwilliams reviewed (approved):
LGTM

I would also pull in the latest master just to be sure there are no conflicts. In the frob page their should be an option to trigger another run on the g3fix page

andrewkolos committed:
fix

andrewkolos committed:
fix nullcheck

andrewkolos committed:
remove redundant error handling

andrewkolos committed:
make naming more consistent

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
add comments

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
Merge branch 'master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
upgrade dependencies

jonahwilliams commented on dev/benchmarks/complex_layout/pubspec.yaml:
```diff
@@ -16,16 +16,18 @@ dependencies:
   flutter_gallery_assets: 1.0.2
 
-  archive: 3.3.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  archive: 3.3.4 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```
It looks like archive added pointycastle as a dependency. Until we decide what to do here, I would recommend pinning the version and filing a bug

christopherfujino replied:
Agreed; I was planning to investigate this tomorrow.

jonahwilliams commented on dev/benchmarks/complex_layout/pubspec.yaml:
```diff
@@ -16,16 +16,18 @@ dependencies:
+  archive: 3.3.4 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   characters: 1.2.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   collection: 1.17.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
+  convert: 3.1.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```
And this one is a dependency of pointycastle

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

christopherfujino commented:
Looks like after merging, this branch is still upgrading package:archive past the pin I added. I would suggest reverting all the pubspec.yaml changes from this branch relate to upstream, and then re-adding the new dependency and then re-running: `flutter update-package --force-upgrade`.

andrewkolos committed:
hmm

andrewkolos committed:
restore collection as prod dependency

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
trying something out

andrewkolos committed:
lint

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into add-binary-encoding-for-asset-manifest

andrewkolos committed:
fix json string parsing

andrewkolos committed:
add test for AssetManifest.json

andrewkolos committed:
remove dependency on collection

andrewkolos committed:
add a comment to hacky try-catch

andrewkolos committed:
restore collection as dev dependency

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -225,16 +251,69 @@ abstract class CachingAssetBundle extends AssetBundle {
+        result = SynchronousFuture<T>(value);
+        if (completer != null) {
+          // The load and parse operation ran asynchronously. We already returned
+          // from the loadStructuredBinaryData function an thus the caller
```
```suggestion
          // from the loadStructuredBinaryData function and thus the caller
```

jonahwilliams commented on packages/flutter/lib/src/services/asset_bundle.dart:
```diff
@@ -225,16 +251,69 @@ abstract class CachingAssetBundle extends AssetBundle {
+        completer!.completeError(err, stack);
+      });
+
+    if (result != null) {
+      // The above code ran synchronously. We can synchronously return the result.
+      _structuredBinaryDataCache[key] = result!;
+      return result!;
+    } else {
```
uber nit, if you return in an `if` block, then don't use an `else` for the other condition

jonahwilliams commented on packages/flutter/test/painting/image_resolution_test.dart:
```diff
@@ -13,18 +13,14 @@ import 'package:flutter_test/flutter_test.dart';
   TestAssetBundle(this._assetBundleMap);
 
-  final Map<String, List<String>> _assetBundleMap;
+  final Map<dynamic, dynamic> _assetBundleMap;
```
even though this is a test, prefer using real types, or at least `Object?`

jonahwilliams reviewed (approved):
LGTM with some minor nits.

Lets hold this until after we cut the release branch

andrewkolos committed:
nits

jonahwilliams commented:
We've branched and this should be safe to land.

andrewkolos commented:
I was hoping FRoBs would be fixed soon so that Google testing could finish, but testing did pass three commits ago, so I will go ahead and land this.

andrewkolos commented:
Manually merged since FRoB is down.

christopherfujino commented:
> SkiaPerf is reporting a nice improvement here: https://flutter-flutter-perf.skia.org/e/?begin=1670172924&end=1670276486&keys=X2c1c806b3584886f3337add1a572b1ea&num_commits=50&request_type=1&xbaroffset=32385

Woohoo!
